File: logistic_regression/logistic_regression.py
import numpy as np
import matplotlib.pyplot as plt

"""Implementation of the logistic regression model."""

class LogisticRegression:
    def __init__(self, learning_rate):
        self.learning_rate = learning_rate
        self.weights = np.random.rand(1, 1)  # Initialize weights randomly
        self.bias = np.random.rand(1)  # Initialize bias randomly
        self.loss_history = []


    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def hypothesis(self, x):
        x = np.dot(x, self.weights) + self.bias
        return self.sigmoid(x)

    def binary_cross_entropy(self, y_true, y_pred):
        epsilon = 1e-15 # to avoid log(0)
        y_pred = np.clip(y_pred, epsilon, 1 - epsilon)  # limit range of predictions
        
        # Binary cross entropy loss
        loss = -(y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred))
        return np.mean(loss)

    def fit(self, x, y, epochs):
        n_samples = x.shape[0]
        y = y.reshape(-1, 1)

        for epoch in range(epochs):
            y_pred = self.hypothesis(x)
            loss = self.binary_cross_entropy(y, y_pred)
            self.loss_history.append(loss)

            error = y_pred - y

            # derivatives
            dW = (1 / n_samples) * np.dot(x.T, error)
            db = (1 / n_samples) * np.sum(error)

            # update weights and bias
            self.weights -= self.learning_rate * dW
            self.bias -= self.learning_rate * db

        return self.loss_history

np.random.seed(42)
X = 2 * np.random.rand(100, 1)
# X가 1을 기준으로 0과 1로 나뉘도록 y 생성
y = (X > 1).astype(int)

model = LogisticRegression(learning_rate=0.1)
losses = model.fit(X, y, epochs=1000)

plt.style.use('seaborn-v0_8-whitegrid')
fig, axes = plt.subplots(1, 2, figsize=(16, 6))
fig.suptitle('Logistic Regression Model Analysis', fontsize=16)

ax1 = axes[0]
ax1.plot(losses, color='dodgerblue', linewidth=2)
ax1.set_title('BCE Loss over Epochs', fontsize=14)
ax1.set_xlabel('Epochs')
ax1.set_ylabel('Binary Cross-Entropy Loss')

ax2 = axes[1]
ax2.scatter(X[y==0], y[y==0], alpha=0.7, edgecolors='k', label='Class 0')
ax2.scatter(X[y==1], y[y==1], alpha=0.7, edgecolors='k', label='Class 1')

x_boundary = np.linspace(X.min(), X.max(), 100).reshape(-1, 1)
y_boundary = model.hypothesis(x_boundary)
ax2.plot(x_boundary, y_boundary, color='crimson', linewidth=3, label='Predicted Probability')
ax2.axhline(y=0.5, color='gray', linestyle='--', label='Threshold (0.5)') # 0.5 기준선

ax2.set_title('Logistic Regression Fit', fontsize=14)
ax2.set_xlabel('X (Feature)')
ax2.set_ylabel('Probability (y_pred)')
ax2.legend()

plt.tight_layout(rect=[0, 0, 1, 0.96])
plt.savefig("logistic_regression_analysis_corrected.png", dpi=150)
plt.show()
plt.close()

File: logistic_regression/cpp/LogisticRegression.cpp
#include "LogisticRegression.hpp"

LogisticRegression::LogisticRegression() : a(0.1) {}

LogisticRegression::~LogisticRegression() {}

double LogisticRegression::sigmoid(double z) const {
    return 1.0 / (1.0 + std::exp(-z));
}

double LogisticRegression::hypothesis(double x) const {
    return sigmoid(w * x + b);
}

// Just computing binary cross entropy, I will use it after.
// currently not using in this codes. Keep it now
double LogisticRegression::binaryCrossEntropy(const std::vector<int>& y_true, const std::vector<double>& y_pred) const {
    if (y_true.size() != y_pred.size()) {
        throw std::invalid_argument("Vectors must be of the same size for binary cross-entropy calculation.");
    }

    double loss = 0.0;
    for (size_t i = 0; i < y_true.size(); ++i) {
        double eps = 1e-15; // to avoid log(0)
        double clipped_pred = std::max(eps, std::min(1 - eps, y_pred[i]));
        loss += -(y_true[i] * std::log(clipped_pred) + (1 - y_true[i]) * std::log(1 - clipped_pred));
    }

    double meanLoss = loss / y_true.size();
    return meanLoss;
}

void LogisticRegression::fit(const std::vector<double>& X, const std::vector<int>& y, int epochs, std::vector<double>& lossHistory) {
    if( X.size() != y.size()) {
        throw std::invalid_argument("Input and output vectors must be of the same size.");
    }
    
    w = 0.0;
    b = 0.0;

    const double n = static_cast<double>(X.size());

    for (int epoch = 0; epoch < epochs; ++epoch) {
        double dWSum = 0.0;
        double dbSum = 0.0;
        double currentLoss = 0.0;

        for (size_t i = 0; i < n; ++i) {
            double yPred = hypothesis(X[i]);
            double error = yPred - y[i];

            dWSum += error * X[i];
            dbSum += error;

            // Calculate current loss with binary cross entropy
            double eps = 1e-15; // to avoid log(0)
            double clipped_pred = std::max(eps, std::min(1 - eps, yPred));
            currentLoss += -(y[i] * std::log(clipped_pred) + (1 - y[i]) * std::log(1.0 - clipped_pred));
        }
        
        // Update weights and bias
        dWSum /= n;
        dbSum /= n;
        
        
        w -= a * dWSum;
        b -= a * dbSum;
        
        lossHistory.push_back(currentLoss / n);
    }
}


int main() {
    std::vector<double> X;
    std::vector<int> y;
    std::default_random_engine generator(42);
    std::uniform_real_distribution<double> distribution(0.0, 2.0);
    
    for (int i = 0; i < 100; ++i) {
        double x_val = distribution(generator);
        X.push_back(x_val);
        y.push_back(x_val > 1.0 ? 1 : 0);
    }

    LogisticRegression model;
    std::vector<double> lossHistory;
    model.fit(X, y, 1000, lossHistory);

    std::ofstream results_file("results.csv");
    results_file << "X,y_true\n";
    for(size_t i = 0; i < X.size(); ++i) {
        results_file << X[i] << "," << y[i] << "\n";
    }
    results_file.close();

    std::ofstream boundary_file("boundary.csv");
    boundary_file << "x_boundary,y_boundary\n";
    for(int i=0; i<=100; ++i) {
        double x_val = 2.0 * i / 100.0;
        boundary_file << x_val << "," << model.hypothesis(x_val) << "\n";
    }
    boundary_file.close();

    std::ofstream loss_file("loss_history.csv");
    loss_file << "loss\n";
    for(const auto& l : lossHistory) {
        loss_file << l << "\n";
    }
    loss_file.close();
    
    std::cout << "C++ Logistic Regression training finished. Results saved to CSV files." << std::endl;
    return 0;
}